File: queries/h07.py
'Volume Shipping Query (Q7)'

from utils import add_date


def tpc_h07(con, NATION1="FRANCE", NATION2="GERMANY", DATE='1995-01-01'):
    supplier = con.table('supplier')
    lineitem = con.table('lineitem')
    orders = con.table('orders')
    customer = con.table('customer')
    nation = con.table('nation')

    q = supplier
    q = q.join(lineitem, supplier.S_SUPPKEY == lineitem.L_SUPPKEY)
    q = q.join(orders, orders.O_ORDERKEY == lineitem.L_ORDERKEY)
    q = q.join(customer, customer.C_CUSTKEY == orders.O_CUSTKEY)
    n1 = nation
    n2 = nation.view()
    q = q.join(n1, supplier.S_NATIONKEY == n1.N_NATIONKEY)
    q = q.join(n2, customer.C_NATIONKEY == n2.N_NATIONKEY)

    q = q[
            n1.N_NAME.name("supp_nation"),
            n2.N_NAME.name("cust_nation"),
            lineitem.L_SHIPDATE,
            lineitem.L_EXTENDEDPRICE,
            lineitem.L_DISCOUNT,
            lineitem.L_SHIPDATE.year().cast('string').name('l_year'),
            (lineitem.L_EXTENDEDPRICE*(1-lineitem.L_DISCOUNT)).name('volume')
    ]

    q = q.filter([
        ((q.cust_nation == NATION1) & (q.supp_nation == NATION2)) |
        ((q.cust_nation == NATION2) & (q.supp_nation == NATION1)),
        q.L_SHIPDATE.between(DATE, add_date(DATE, dy=2, dd=-1))
        ])

    gq = q.group_by(['supp_nation', 'cust_nation', 'l_year'])
    q = gq.aggregate(revenue=q.volume.sum())
    q = q.sort_by(['supp_nation', 'cust_nation', 'l_year'])

    return q

File: dplyr_tpc/h07.R
tpc_h07 <- function(input_func, collect_func = dplyr::collect) {
  sn <- inner_join(
    input_func("supplier") %>%
      select(s_nationkey, s_suppkey),
    input_func("nation") %>%
      select(n1_nationkey = n_nationkey, n1_name = n_name) %>%
      filter(n1_name %in% c("FRANCE", "GERMANY")),
    by = c("s_nationkey" = "n1_nationkey")) %>%
    select(s_suppkey, n1_name)

  cn <- inner_join(
    input_func("customer") %>%
      select(c_custkey, c_nationkey),
    input_func("nation") %>%
      select(n2_nationkey = n_nationkey, n2_name = n_name) %>%
      filter(n2_name %in% c("FRANCE", "GERMANY")),
    by = c("c_nationkey" = "n2_nationkey")) %>%
    select(c_custkey, n2_name)

  cno <- inner_join(
    input_func("orders") %>%
      select(o_custkey, o_orderkey),
    cn, by = c("o_custkey" = "c_custkey")) %>%
    select(o_orderkey, n2_name)

  cnol <- inner_join(
    input_func("lineitem") %>%
      select(l_orderkey, l_suppkey, l_shipdate, l_extendedprice, l_discount) %>%
      # kludge, should be: filter(l_shipdate >= "1995-01-01", l_shipdate <= "1996-12-31"),
      filter(l_shipdate >= as.Date("1995-01-01"), l_shipdate <= as.Date("1996-12-31")),
    cno,
    by = c("l_orderkey" = "o_orderkey")) %>%
    select(l_suppkey, l_shipdate, l_extendedprice, l_discount, n2_name)

  all <- inner_join(cnol, sn, by = c("l_suppkey" = "s_suppkey"))

  all %>%
    filter((n1_name == "FRANCE" & n2_name == "GERMANY") |
             (n1_name == "GERMANY" & n2_name == "FRANCE")) %>%
    mutate(
      supp_nation = n1_name,
      cust_nation = n2_name,
      # kludge (?) l_year = as.integer(strftime(l_shipdate, "%Y")),
      l_year = strftime(l_shipdate, "%Y"), # year(l_shipdate),
      volume = l_extendedprice * (1 - l_discount)) %>%
    select(supp_nation, cust_nation, l_year, volume) %>%
    group_by(supp_nation, cust_nation, l_year) %>%
    summarise(revenue = sum(volume)) %>%
    ungroup() %>%
    arrange(supp_nation, cust_nation, l_year) %>%
    collect_func()
}
